internal/core/src/index/IndexFactory.cpp

```diff
@@ -142,47 +142,90 @@ IndexFactory::CreatePrimitiveScalarIndex(
     switch (data_type) {
         // create scalar index
         case DataType::BOOL:
-            return CreateScalarIndex<bool>(index_type, file_manager_context);
+            return CreatePrimitiveScalarIndex<bool>(index_type,
+                                                    file_manager_context);
         case DataType::INT8:
-            return CreateScalarIndex<int8_t>(index_type, file_manager_context);
+            return CreatePrimitiveScalarIndex<int8_t>(index_type,
+                                                      file_manager_context);
         case DataType::INT16:
-            return CreateScalarIndex<int16_t>(index_type, file_manager_context);
+            return CreatePrimitiveScalarIndex<int16_t>(index_type,
+                                                       file_manager_context);
         case DataType::INT32:
-            return CreateScalarIndex<int32_t>(index_type, file_manager_context);
+            return CreatePrimitiveScalarIndex<int32_t>(index_type,
+                                                       file_manager_context);
         case DataType::INT64:
-            return CreateScalarIndex<int64_t>(index_type, file_manager_context);
+            return CreatePrimitiveScalarIndex<int64_t>(index_type,
+                                                       file_manager_context);
         case DataType::FLOAT:
-            return CreateScalarIndex<float>(index_type, file_manager_context);
+            return CreatePrimitiveScalarIndex<float>(index_type,
+                                                     file_manager_context);
         case DataType::DOUBLE:
-            return CreateScalarIndex<double>(index_type, file_manager_context);
+            return CreatePrimitiveScalarIndex<double>(index_type,
+                                                      file_manager_context);
 
             // create string index
         case DataType::STRING:
         case DataType::VARCHAR:
-            return CreateScalarIndex<std::string>(index_type,
-                                                  file_manager_context);
+            return CreatePrimitiveScalarIndex<std::string>(
+                index_type, file_manager_context);
         default:
             throw SegcoreError(
                 DataTypeInvalid,
                 fmt::format("invalid data type to build index: {}", data_type));
     }
 }
 
+IndexBasePtr
+IndexFactory::CreateCompositeScalarIndex(
+    IndexType index_type,
+    const storage::FileManagerContext& file_manager_context) {
+    if (index_type == BITMAP_INDEX_TYPE) {
+        auto element_type = static_cast<DataType>(
+            file_manager_context.fieldDataMeta.field_schema.element_type());
+        return CreatePrimitiveScalarIndex(
+            element_type, index_type, file_manager_context);
+    } else if (index_type == INVERTED_INDEX_TYPE) {
```

Comment:
A little duplicate. I think we can return CreatePrimitiveScalarIndex directly, since the index type will be dispatched inner CreatePrimitiveScalarIndex.